proto/substrait/algebra.proto

```diff
@@ -173,6 +173,8 @@ message JoinRel {
     // This join is useful for nested sub-queries where we need exactly one record in output (or throw exception)
     // See Section 3.2 of https://15721.courses.cs.cmu.edu/spring2018/papers/16-optimizer2/hyperjoins-btw2017.pdf
     JOIN_TYPE_SINGLE = 7;
+    JOIN_TYPE_MARK = 8;
```

Comment:
I put in a PR to align the join types to make this PR simpler.  I'd like to separate out adding the mark join type as well.  https://github.com/substrait-io/substrait/pull/663 could be the MARK join PR but the documentation would need updating.  From the comments on https://github.com/substrait-io/substrait/pull/625/files#r1589078241 it sounds like MarkJoin isn't actually a join but a specialized projection.  As such it doesn't follow the normal field output rules for a JoinRel.  It could be implemented as a new relation type but it could also just be a projection with a hint reminding us that it's just a mark relation.

Reply:
I'm not sure I would agree with this; a mark join does not alter the cardinality of the LHS. However, its result on the mark column depends on the result of the join condition. For example, from an implementation perspective, the mark can be viewed as a specialization of a hash join.

Reply:
From your comment in that PR looked like it emitted three columns per row including the mark.  But if that's an internal detail there is no issue with it being another kind of hash join.